File: Julia/CalculatorTests.jl
#:----------------------------------------------------------------------------
#:
#:  Copyright © 2023 Steffen Liersch
#:  https://www.steffen-liersch.de/
#:
#:----------------------------------------------------------------------------

include("Calculator.jl")
include("FloatFormatter.jl")
include("Helpers.jl")

function performtests(tests)::Bool
  testcount = 0
  errorcount = 0
  calculator = Calculator(FloatFormatter("%.16G"))
  for test in tests
    println("Test: " * test["name"])
    array = isa(test["expression"], Array) ? test["expression"] : [test["expression"]]
    for expr in array
      lines = calculateandformat!(calculator, expr, "")
      if !assertequal(test["output"], lines)
        errorcount += 1
      end
      testcount += 1
    end
  end
  println("=> $testcount tests completed with $errorcount errors")
  return errorcount <= 0
end

function assertequal(expected::Vector, actual::Vector)::Bool
  c1 = length(expected)
  c2 = length(actual)
  if c1 != c2
    println("Assertion failed: Arrays with different lengths")
    return false
  end

  for i in 1:c1
    if expected[i] != actual[i]
      println("Assertion failed: $(expected[i]) != $(actual[i])")
      return false
    end
  end

  return true
end

File: JavaScript/calculator-tests.js
/*--------------------------------------------------------------------------*\
::
::  Copyright © 2023 Steffen Liersch
::  https://www.steffen-liersch.de/
::
\*--------------------------------------------------------------------------*/

export { performTests };
import { Calculator } from './calculator.js';
import { FloatFormatter } from './float-formatter.js';

function performTests(tests)
{
  let testCount = 0;
  let errorCount = 0;
  let calculator = new Calculator(new FloatFormatter(16));
  for(let test of tests)
  {
    console.log('Test: ' + test.name);
    let array = Array.isArray(test.expression) ? test.expression : [test.expression];
    for(let expr of array)
    {
      let lines = calculator.calculateAndFormat(expr, '');
      if(!assertEqual(test.output, lines))
        errorCount++;
      testCount++;
    }
  }
  console.log(`=> ${testCount} tests completed with ${errorCount} errors`);
  return errorCount <= 0;
}

function assertEqual(expected, actual)
{
  if(!Array.isArray(expected)) throw Error('Array expected (1)');
  if(!Array.isArray(actual)) throw Error('Array expected (2)');

  let c1 = expected.length;
  let c2 = actual.length;
  if(c1 != c2)
  {
    console.error('Assertion failed: Arrays with different lengths', expected, actual);
    return false;
  }

  for(let i = 0; i < c1; i++)
  {
    if(expected[i] != actual[i])
    {
      console.error('Assertion failed', expected, actual);
      return false;
    }
  }

  return true;
}
